Support bup 0.34, while also still working with previous versions.

bup 0.34 no longer allows `-r` with a plain path, only host:path and some URI
forms. The fix is to use `bup -d` when the buprepo is a path, which also works
with older bup. Without this, the bup special remote is broken against the
bup 0.34 that nixpkgs currently pins.

Upstream commit: 0b0bd012976b20b3857d532f00a1fc7a9bab12a3
Author: Joey Hess <joeyh@joeyh.name>
Date: Sun, 27 Sep 2026 15:23:59 -0400
Upstream bug: https://git-annex.branchable.com/bugs/support_new_bup_version/

REBASED onto git-annex 10.20260421 (the version nixpkgs pins). Upstream wrote
this against the unreleased 10.20260902 tree, so it does not apply verbatim:

  * Two Remote/Bup.hs hunks (bupParams, bup2GitRemote) missed only because
    upstream stripped trailing whitespace from those lines after 10.20260421.
  * The CHANGELOG hunk targets a "10.20260902 UNRELEASED" section that does
    not exist here, and doc/bugs/support_new_bup_version.mdwn does not exist
    at this tag at all. Both are cosmetic and are dropped.

Only the functional Remote/Bup.hs change is kept, regenerated so it applies
at plain -p1 with no fuzz. Drop this patch once nixpkgs' git-annex reaches
10.20260902 or later, where it is included upstream.

--- a/Remote/Bup.hs
+++ b/Remote/Bup.hs
@@ -1,6 +1,6 @@
 {- Using bup as a remote.
  -
- - Copyright 2011-2022 Joey Hess <id@joeyh.name>
+ - Copyright 2011-2026 Joey Hess <id@joeyh.name>
  -
  - Licensed under the GNU AGPL version 3 or higher.
  -}
@@ -41,7 +41,22 @@
 import Utility.Metered
 import Types.ProposedAccepted
 
-type BupRepo = String
+data BupRepo
+	= BupRepoPath FilePath
+	| BupRepoRemote String
+
+parseBupRepo :: String -> BupRepo
+parseBupRepo s
+	| ':' `elem` s = BupRepoRemote s
+	| otherwise = BupRepoPath s
+
+serializeBupRepo :: BupRepo -> String
+serializeBupRepo (BupRepoPath p) = p
+serializeBupRepo (BupRepoRemote r) = r
+
+bupRepoLocal :: BupRepo -> Bool
+bupRepoLocal (BupRepoPath _) = True
+bupRepoLocal (BupRepoRemote _ ) = False
 
 remote :: RemoteType
 remote = specialRemoteType $ RemoteType
@@ -66,7 +81,7 @@
 	c <- parsedRemoteConfig remote rc
 	bupr <- liftIO $ bup2GitRemote buprepo
 	cst <- remoteCost gc c $
-		if bupLocal buprepo
+		if bupRepoLocal buprepo
 			then nearlyCheapRemoteCost
 			else expensiveRemoteCost
 	(u', bupr') <- getBupUUID bupr u
@@ -85,7 +100,7 @@
 		, removeKey = removeKeyDummy
 		, lockContent = Nothing
 		, checkPresent = checkPresentDummy
-		, checkPresentCheap = bupLocal buprepo
+		, checkPresentCheap = bupRepoLocal buprepo
 		, exportActions = exportUnsupported
 		, importActions = importUnsupported
 		, whereisKey = Nothing
@@ -95,18 +110,21 @@
 		, config = c
 		, getRepo = return r
 		, gitconfig = gc
-		, localpath = if bupLocal buprepo && not (null buprepo)
-			then Just (toOsPath buprepo)
-			else Nothing
+		, localpath = case buprepo of
+			BupRepoPath p | not (null p) -> Just (toOsPath p)
+			_ -> Nothing
 		, remotetype = remote
-		, availability = if null buprepo
-			then pure LocallyAvailable
-			else checkPathAvailability (bupLocal buprepo) (toOsPath buprepo)
+		, availability = case buprepo of
+			BupRepoPath p
+				| null p -> pure LocallyAvailable
+				| otherwise ->
+					checkPathAvailability True (toOsPath p)
+			BupRepoRemote _ -> pure GloballyAvailable
 		, readonly = False
 		, appendonly = False
 		, untrustworthy = False
 		, mkUnavailable = return Nothing
-		, getInfo = return [("repo", buprepo)]
+		, getInfo = return [("repo", serializeBupRepo buprepo)]
 		, claimUrl = Nothing
 		, checkUrl = Nothing
 		, remoteStateHandle = rs
@@ -122,15 +140,17 @@
 		(checkKey bupr')
 		this
   where
-	buprepo = fromMaybe (giveup "missing buprepo") $ remoteAnnexBupRepo gc
+	buprepo = maybe (giveup "missing buprepo") parseBupRepo $
+		remoteAnnexBupRepo gc
 
 bupSetup :: SetupStage -> Maybe UUID -> RemoteName -> Maybe CredPair -> RemoteConfig -> RemoteGitConfig -> Annex (RemoteConfig, UUID)
 bupSetup ss mu _ _ c gc = do
 	u <- maybe (liftIO genUUID) return mu
 
 	-- verify configuration is sane
-	let buprepo = maybe (giveup "Specify buprepo=") fromProposedAccepted $
-		M.lookup buprepoField c
+	let buprepo = maybe (giveup "Specify buprepo=")
+		(parseBupRepo . fromProposedAccepted)
+		(M.lookup buprepoField c)
 	(c', _encsetup) <- encryptionSetup ss c gc
 
 	-- bup init will create the repository.
@@ -142,13 +162,15 @@
 
 	-- The buprepo is stored in git config, as well as this repo's
 	-- persistent state, so it can vary between hosts.
-	gitConfigSpecialRemote u c' [("buprepo", buprepo)]
+	gitConfigSpecialRemote u c' [("buprepo", serializeBupRepo buprepo)]
 
 	return (c', u)
 
 bupParams :: String -> BupRepo -> [CommandParam] -> [CommandParam]
-bupParams command buprepo params = 
-	Param command : [Param "-r", Param buprepo] ++ params
+bupParams command (BupRepoPath buprepo) params =
+	Param "-d" : Param buprepo : Param command : params
+bupParams command (BupRepoRemote buprepo) params =
+	Param command : Param "-r" : Param buprepo : params
 
 bup :: String -> BupRepo -> [CommandParam] -> Annex Bool
 bup command buprepo params = do
@@ -293,19 +315,18 @@
 			Right r' -> return (toUUID $ Git.Config.get configkeyUUID mempty r', r')
 			Left _ -> return (NoUUID, r)
 
-{- Converts a bup remote path spec into a Git.Repo. There are some
- - differences in path representation between git and bup. -}
+{- Converts a BupRepo into a Git.Repo. There are some
+ - differences in representation between git and bup. -}
 bup2GitRemote :: BupRepo -> IO Git.Repo
-bup2GitRemote "" = do
-	-- bup -r "" operates on ~/.bup
+bup2GitRemote (BupRepoPath "") = do
+	-- bup -d "" operates on ~/.bup
 	h <- myHomeDir
 	Git.Construct.fromPath $ toOsPath h </> literalOsPath ".bup"
-bup2GitRemote r
-	| bupLocal r = 
-		if "/" `isPrefixOf` r
-			then Git.Construct.fromPath (toOsPath r)
-			else giveup "please specify an absolute path"
-	| otherwise = Git.Construct.fromUrl False $ "ssh://" ++ host ++ slash dir
+bup2GitRemote (BupRepoPath r)
+	| "/" `isPrefixOf` r = Git.Construct.fromPath (toOsPath r)
+	| otherwise = giveup "please specify an absolute path"
+bup2GitRemote (BupRepoRemote r) =
+	Git.Construct.fromUrl False $ "ssh://" ++ host ++ slash dir
   where
 	bits = splitc ':' r
 	host = fromMaybe "" $ headMaybe bits
@@ -327,9 +348,6 @@
   where
 	shown = serializeKey k
 
-bupLocal :: BupRepo -> Bool
-bupLocal = notElem ':'
-
 {- Bup is not concurrency safe, so use a lock file. Only one writer process
  - should run at a time; multiple readers may run if no writer is running. -}
 lockBup :: Bool -> Remote -> Annex a -> Annex a
